rts/motoko-rts/src/gc/generational.rs

```diff
@@ -0,0 +1,490 @@
+        };
+        self.heap.limits.free - base
+    }
+
+    fn survival_rate(&self) -> f64 {
+        self.marked_space as f64 / self.generation_size() as f64
+    }
+
+    unsafe fn mark_phase(&mut self) {
+        self.marked_space = 0;
+        self.mark_root_set();
+        self.mark_all_reachable();
+    }
+
+    unsafe fn mark_root_set(&mut self) {
+        self.mark_static_roots();
+
+        if (*self.heap.roots.continuation_table_ptr_loc).is_ptr() {
+            self.mark_object(*self.heap.roots.continuation_table_ptr_loc);
+        }
+
+        if self.strategy == Strategy::Young {
+            self.mark_additional_young_root_set();
+        }
+    }
+
+    unsafe fn mark_static_roots(&mut self) {
+        let root_array = self.heap.roots.static_roots.as_array();
+        for i in 0..root_array.len() {
+            let object = root_array.get(i).as_obj();
+            debug_assert_eq!(object.tag(), TAG_MUTBOX);
+            debug_assert!((object as usize) < self.heap.limits.base);
+            self.mark_root_mutbox_fields(object as *mut MutBox);
+        }
+    }
+
+    unsafe fn mark_additional_young_root_set(&mut self) {
+        let mut iterator = REMEMBERED_SET.as_ref().unwrap().iterate();
+        while iterator.has_next() {
+            let location = iterator.current().get_raw() as *mut Value;
+            let value = *location;
+            // Check whether the location still refers to young object as this may have changed
+            // due to subsequent writes to that location after the write barrier recording.
+            if value.is_ptr() && (value.get_raw() as usize) >= self.heap.limits.last_free {
+                self.mark_object(value);
+            }
+            iterator.next();
+        }
+    }
+
+    unsafe fn mark_object(&mut self, object: Value) {
+        let pointer = object.get_ptr() as u32;
+        if self.strategy == Strategy::Young && pointer < self.heap.limits.last_free as u32 {
+            return;
+        }
+        debug_assert_eq!(pointer % WORD_SIZE, 0);
+
+        let obj_idx = pointer / WORD_SIZE;
+        if get_bit(obj_idx) {
+            return;
+        }
+        set_bit(obj_idx);
+
+        push_mark_stack(self.heap.mem, pointer as usize, object.tag());
+        self.marked_space += object_size(pointer as usize).to_bytes().as_usize();
+    }
+
+    unsafe fn mark_all_reachable(&mut self) {
+        while let Some((obj, tag)) = pop_mark_stack() {
+            self.mark_fields(obj as *mut Obj, tag)
+        }
+    }
+
+    unsafe fn mark_fields(&mut self, object: *mut Obj, object_tag: Tag) {
+        visit_pointer_fields(
+            self,
+            object,
+            object_tag,
+            self.heap.limits.base,
+            |gc, field_address| {
+                let field_value = *field_address;
+                gc.mark_object(field_value);
+            },
+            |gc, slice_start, array| {
+                const SLICE_INCREMENT: u32 = 127;
```

Comment:
Maybe we should define SLICE_INCREMENT more prominently somewhere

Reply:
The constant is not really a relevant one, but only for scalability tuning inside the GC, such that the breadth-first search does not copy too many element pointers in the mark stack at once. I copied that logic from the existing compacting GC. The constant could be locally changed without any impact on the remaining logic or other GCs.

Reply:
Ok